File: backend/app/core/kinematics.py
"""
Core kinematics module.

Provides a vectorized, numpy-based forward-kinematics and Jacobian
implementation for serial revolute manipulators defined via standard
Denavit-Hartenberg (DH) parameters. Every solver (classical or
protein-inspired) is built on top of this shared, efficiency-minded core
so that comparisons between solvers are apples-to-apples and fast.

Design notes:
- All transforms are computed with a single vectorized chain multiply
  rather than per-joint Python-level objects, since this is the hottest
  path in every solver (called every iteration, often thousands of times
  per solve, and millions of times across a benchmark sweep).
- Joint limits, link geometry (for self-collision), and DH parameters are
  bundled into an immutable `RobotSpec` so every solver/benchmark consumes
  exactly the same robot definition.
"""

from __future__ import annotations

import math
import numpy as np
from dataclasses import dataclass, field


@dataclass(frozen=True)
class RobotSpec:
    """Defines a serial revolute manipulator via standard DH parameters.

    DH convention used (standard/Denavit, not modified):
        T_i = Rot_z(theta_i) * Trans_z(d_i) * Trans_x(a_i) * Rot_x(alpha_i)

    Attributes:
        name: human readable id, e.g. "ur5"
        a: link lengths (n,)
        d: link offsets (n,)
        alpha: link twists (n,)
        theta_offset: fixed offset added to each joint variable (n,)
        joint_limits: (n, 2) array of [lower, upper] radians per joint
        link_radius: approximate cylindrical radius per link, used for
            cheap self-collision / "steric clash" distance checks (n,)
    """

    name: str
    a: np.ndarray
    d: np.ndarray
    alpha: np.ndarray
    theta_offset: np.ndarray
    joint_limits: np.ndarray
    link_radius: np.ndarray

    @property
    def n_joints(self) -> int:
        return len(self.a)

    def random_config(self, rng: np.random.Generator) -> np.ndarray:
        lo = self.joint_limits[:, 0]
        hi = self.joint_limits[:, 1]
        return rng.uniform(lo, hi)

    def clip(self, q: np.ndarray) -> np.ndarray:
        return np.clip(q, self.joint_limits[:, 0], self.joint_limits[:, 1])


def ur5_spec() -> RobotSpec:
    """Standard UR5 DH parameters (meters, radians).

    Source values are the well-known UR5 DH table (Universal Robots),
    used widely as a benchmark arm in the IK literature.
    """
    a = np.array([0.0, -0.42500, -0.39225, 0.0, 0.0, 0.0])
    d = np.array([0.089159, 0.0, 0.0, 0.10915, 0.09465, 0.0823])
    alpha = np.array([np.pi / 2, 0.0, 0.0, np.pi / 2, -np.pi / 2, 0.0])
    theta_offset = np.zeros(6)
    limit = np.array([2 * np.pi] * 6)  # effectively unlimited but bounded
    joint_limits = np.stack([-limit, limit], axis=1)
    # rough visual/collision radius per link (meters)
    link_radius = np.array([0.06, 0.05, 0.045, 0.04, 0.04, 0.035])
    return RobotSpec(
        name="ur5",
        a=a, d=d, alpha=alpha, theta_offset=theta_offset,
        joint_limits=joint_limits, link_radius=link_radius,
    )


def franka_panda_spec() -> RobotSpec:
    """Franka Emika Panda — 7-DOF robot arm.

    Standard DH parameters (same convention used throughout: T_i = Rot_z * Trans_z * Trans_x * Rot_x).
    Source: Franka documentation + Gaz et al. 2019.

    Notable: joint 4 is permanently in the negative range [-3.07, -0.07] rad
    (the 'elbow-down' kinematic constraint). random_config() and clip() already
    respect this via joint_limits, so all solvers handle it correctly without
    any solver-specific changes. The total workspace reach is ~0.855m.
    """
    a     = np.array([0,       0,       0,       0.0825, -0.0825, 0,      0.088])
    d     = np.array([0.333,   0,       0.316,   0,       0.384,  0,      0.107])
    alpha = np.array([0,      -np.pi/2, np.pi/2, np.pi/2,-np.pi/2, np.pi/2, np.pi/2])
    theta_offset = np.zeros(7)
    joint_limits = np.array([
        [-2.8973,  2.8973],   # q1
        [-1.7628,  1.7628],   # q2
        [-2.8973,  2.8973],   # q3
        [-3.0718, -0.0698],   # q4 — always negative (Franka elbow-down constraint)
        [-2.8973,  2.8973],   # q5
        [-0.0175,  3.7525],   # q6
        [-2.8973,  2.8973],   # q7
    ])
    link_radius = np.array([0.08, 0.07, 0.07, 0.06, 0.06, 0.05, 0.04])
    return RobotSpec(
        name="franka_panda",
        a=a, d=d, alpha=alpha, theta_offset=theta_offset,
        joint_limits=joint_limits, link_radius=link_radius,
    )


def planar3dof_spec() -> RobotSpec:
    """Planar 3-DOF arm (RRR) in the XY plane.

    DH parameters for a standard 3-link planar arm:
        - All joints revolute, rotating about Z.
        - alpha = 0 everywhere (no twist — stays in the XY plane).
        - d = 0 everywhere (no Z offset).
        - Link lengths: L1=0.4m, L2=0.3m, L3=0.2m (total reach 0.9m).

    This arm has a closed-form analytical IK solution, which makes it
    the ideal ground-truth validator: we can compare any numerical solver's
    output against the exact answer, not just check 'error < tolerance'.

    End-effector state is (x, y, theta) where theta = q1+q2+q3 is the
    tip orientation in the plane. The DH FK produces a full 4x4 matrix
    where the relevant state is in the XY block and Z is always ~0.
    All existing FK / Jacobian / pose_error / self-collision machinery
    works without modification since DH is fully general.
    """
    L1, L2, L3 = 0.4, 0.3, 0.2
    a = np.array([L1, L2, L3])
    d = np.zeros(3)
    alpha = np.zeros(3)          # planar — no twist
    theta_offset = np.zeros(3)
    joint_limits = np.array([[-np.pi, np.pi]] * 3)
    # link radii for the thin planar arm (used in self-collision checks)
    link_radius = np.array([0.03, 0.025, 0.02])
    return RobotSpec(
        name="planar3dof",
        a=a, d=d, alpha=alpha, theta_offset=theta_offset,
        joint_limits=joint_limits, link_radius=link_radius,
    )


# ---------------------------------------------------------------------------
# Robot registry — maps robot name → spec factory function.
# Add new arms here; the API will expose them automatically.
# ---------------------------------------------------------------------------
ROBOT_REGISTRY: dict[str, callable] = {
    "ur5":           ur5_spec,
    "planar3dof":    planar3dof_spec,
    "franka_panda":  franka_panda_spec,
}


def get_robot_spec(name: str) -> RobotSpec:
    """Return a fresh RobotSpec for the named robot.

    Raises ValueError for unknown names so callers get a clean message
    rather than a KeyError traceback.
    """
    if name not in ROBOT_REGISTRY:
        raise ValueError(
            f"Unknown robot '{name}'. Available: {list(ROBOT_REGISTRY)}"
        )
    return ROBOT_REGISTRY[name]()


def _dh_transform(theta: float, d: float, a: float, alpha: float) -> np.ndarray:
    ct, st = np.cos(theta), np.sin(theta)
    ca, sa = np.cos(alpha), np.sin(alpha)
    return np.array([
        [ct, -st * ca,  st * sa, a * ct],
        [st,  ct * ca, -ct * sa, a * st],
        [0.0,      sa,       ca,      d],
        [0.0,     0.0,      0.0,    1.0],
    ])


def forward_kinematics_chain(spec: RobotSpec, q: np.ndarray) -> np.ndarray:
    """Returns the full chain of joint-origin transforms.

    Returns:
        (n+1, 4, 4) array: T[0] is base frame (identity), T[i] is the
        transform of joint i's origin in base frame, T[n] is the
        end-effector frame. Computing the whole chain (not just the
        end-effector) is required for: link-midpoint self-collision
        checks, the neighbor/local-interaction energy terms in the
        protein-IK solver, and CCD/FABRIK which both need every joint
        position, not just the tip.
    """
    n = spec.n_joints
    thetas = q + spec.theta_offset
    T = np.empty((n + 1, 4, 4))
    T[0] = np.eye(4)
    for i in range(n):
        Ti = _dh_transform(thetas[i], spec.d[i], spec.a[i], spec.alpha[i])
        T[i + 1] = T[i] @ Ti
    return T


def end_effector_pose(spec: RobotSpec, q: np.ndarray) -> np.ndarray:
    """Returns the 4x4 end-effector transform only (cheaper than full chain)."""
    n = spec.n_joints
    thetas = q + spec.theta_offset
    T = np.eye(4)
    for i in range(n):
        T = T @ _dh_transform(thetas[i], spec.d[i], spec.a[i], spec.alpha[i])
    return T


def joint_positions(spec: RobotSpec, q: np.ndarray) -> np.ndarray:
    """Returns (n+1, 3) array of joint origin positions in base frame."""
    chain = forward_kinematics_chain(spec, q)
    return chain[:, :3, 3]


def geometric_jacobian(spec: RobotSpec, q: np.ndarray) -> np.ndarray:
    """Standard 6xN geometric Jacobian (linear velocity, angular velocity)
    for revolute joints, computed from the joint-origin chain.

    J_v_i = z_i x (p_end - p_i)
    J_w_i = z_i
    where z_i is the joint i rotation axis in base frame and p_i its origin.
    """
    n = spec.n_joints
    chain = forward_kinematics_chain(spec, q)
    p_end = chain[n, :3, 3]
    J = np.zeros((6, n))
    for i in range(n):
        z_i = chain[i, :3, 2]
        p_i = chain[i, :3, 3]
        J[:3, i] = np.cross(z_i, p_end - p_i)
        J[3:, i] = z_i
    return J


def pose_error(T_current: np.ndarray, T_target: np.ndarray) -> np.ndarray:
    """6-vector [position_error(3), orientation_error(3)] between two
    homogeneous transforms. Orientation error uses the axis-angle vector
    of the relative rotation (small-angle-friendly, standard for IK).
    """
    p_err = T_target[:3, 3] - T_current[:3, 3]
    R_err = T_target[:3, :3] @ T_current[:3, :3].T
    # axis-angle (rotation vector) from R_err
    cos_theta = np.clip((np.trace(R_err) - 1.0) / 2.0, -1.0, 1.0)
    theta = np.arccos(cos_theta)
    if theta < 1e-8:
        o_err = np.zeros(3)
    else:
        axis = np.array([
            R_err[2, 1] - R_err[1, 2],
            R_err[0, 2] - R_err[2, 0],
            R_err[1, 0] - R_err[0, 1],
        ]) / (2.0 * np.sin(theta))
        o_err = axis * theta
    return np.concatenate([p_err, o_err])


def self_collision_min_distance(spec: RobotSpec, q: np.ndarray) -> float:
    """Cheap proxy for steric clash: minimum surface-to-surface distance
    between non-adjacent links, approximated as capsule (segment) distance
    between consecutive joint-position segments, minus their radii.
    Only checks non-adjacent link pairs (adjacent links always share a
    joint and are not meaningful "collisions").

    Vectorized across all candidate pairs simultaneously (rather than a
    Python double-loop calling a scalar per-pair function), since this is
    the single hottest path in the protein-IK solver's energy evaluation
    (profiling showed it dominating total solve time before
    vectorization: ~56% of wall-clock time at 789 calls/solve).
    """
    chain = forward_kinematics_chain(spec, q)
    return self_collision_min_distance_from_chain(spec, chain)


def self_collision_min_distance_from_chain(spec: RobotSpec, chain: np.ndarray) -> float:
    """Same as `self_collision_min_distance`, but takes an already-computed
    FK chain (as returned by `forward_kinematics_chain`) instead of
    recomputing it. Lets callers that already need the FK chain for other
    purposes (e.g. target-pose error) avoid a second redundant FK pass --
    profiling identified this redundant double-FK-per-energy-evaluation as
    a significant remaining cost after the segment-distance vectorization.

    Implementation note: uses a plain Python/scalar loop rather than numpy
    vectorization across pairs. This looks backwards, but was verified by
    direct profiling: for the small, fixed number of non-adjacent link
    pairs on a 6-DOF arm (10 pairs), numpy's per-call array-construction
    and broadcasting overhead exceeds the cost of the underlying
    arithmetic, making the vectorized version ~2.6x SLOWER than a tight
    scalar loop (224ms vs 86ms per 1000 calls, measured directly). This is
    the opposite of the usual numpy advice, and is specific to this
    problem's tiny fixed size -- worth revisiting if this solver is ever
    extended to high-DOF (15+ joint) chains where the pair count grows
    quadratically and vectorization would likely win again.
    """
    pts = chain[:, :3, 3]
    n_links = spec.n_joints
    if n_links < 3:
        return 1.0  # no non-adjacent pairs possible

    pts_list = pts.tolist()
    radii = spec.link_radius
    min_d = float("inf")
    for i in range(n_links - 1):
        p1 = pts_list[i]
        p2 = pts_list[i + 1]
        ri = radii[i]
        for j in range(i + 2, n_links):
            d = _segment_segment_distance_scalar(p1, p2, pts_list[j], pts_list[j + 1])
            d -= (ri + radii[j])
            if d < min_d:
                min_d = d
    return min_d


def _segment_segment_distance_scalar(p1, p2, p3, p4) -> float:
    """Plain-Python (list/tuple, not numpy array) closest distance between
    two 3D line segments. Same algorithm as `_segment_segment_distance`
    (Ericson, "Real-Time Collision Detection"), reimplemented without
    numpy array overhead -- see `self_collision_min_distance_from_chain`
    docstring for why this matters at this problem's scale."""
    d1x, d1y, d1z = p2[0] - p1[0], p2[1] - p1[1], p2[2] - p1[2]
    d2x, d2y, d2z = p4[0] - p3[0], p4[1] - p3[1], p4[2] - p3[2]
    rx, ry, rz = p1[0] - p3[0], p1[1] - p3[1], p1[2] - p3[2]

    a = d1x * d1x + d1y * d1y + d1z * d1z
    e = d2x * d2x + d2y * d2y + d2z * d2z
    f = d2x * rx + d2y * ry + d2z * rz
    eps = 1e-12

    if a <= eps and e <= eps:
        dx, dy, dz = p1[0] - p3[0], p1[1] - p3[1], p1[2] - p3[2]
        return math.sqrt(dx * dx + dy * dy + dz * dz)

    if a <= eps:
        s = 0.0
        t = min(max(f / e, 0.0), 1.0)
    else:
        c = d1x * rx + d1y * ry + d1z * rz
        if e <= eps:
            t = 0.0
            s = min(max(-c / a, 0.0), 1.0)
        else:
            b = d1x * d2x + d1y * d2y + d1z * d2z
            denom = a * e - b * b
            s = min(max((b * f - c * e) / denom, 0.0), 1.0) if denom > eps else 0.0
            t = (b * s + f) / e
            if t < 0.0:
                t = 0.0
                s = min(max(-c / a, 0.0), 1.0)
            elif t > 1.0:
                t = 1.0
                s = min(max((b - c) / a, 0.0), 1.0)

    c1x, c1y, c1z = p1[0] + d1x * s, p1[1] + d1y * s, p1[2] + d1z * s
    c2x, c2y, c2z = p3[0] + d2x * t, p3[1] + d2y * t, p3[2] + d2z * t
    dx, dy, dz = c1x - c2x, c1y - c2y, c1z - c2z
    return math.sqrt(dx * dx + dy * dy + dz * dz)


def _segment_segment_distance_batch(p1: np.ndarray, p2: np.ndarray,
                                     p3: np.ndarray, p4: np.ndarray) -> np.ndarray:
    """Vectorized closest distance between many pairs of 3D line segments
    at once. Same algorithm as `_segment_segment_distance` (Ericson,
    "Real-Time Collision Detection"), operating on (M,3) arrays of M
    segment pairs simultaneously via numpy broadcasting instead of a
    Python loop calling the scalar version M times.
    """
    d1 = p2 - p1
    d2 = p4 - p3
    r = p1 - p3
    a = np.sum(d1 * d1, axis=1)
    e = np.sum(d2 * d2, axis=1)
    f = np.sum(d2 * r, axis=1)
    c = np.sum(d1 * r, axis=1)
    b = np.sum(d1 * d2, axis=1)

    eps = 1e-12
    denom = a * e - b * b

    s = np.zeros_like(a)
    t = np.zeros_like(a)

    both_degenerate = (a <= eps) & (e <= eps)
    a_degenerate = (a <= eps) & ~both_degenerate
    e_degenerate = (e <= eps) & ~both_degenerate & ~a_degenerate
    general = ~both_degenerate & ~a_degenerate & ~e_degenerate

    # a degenerate (p1==p2): s=0, t = clip(f/e)
    t[a_degenerate] = np.clip(f[a_degenerate] / np.where(e[a_degenerate] > eps, e[a_degenerate], 1.0), 0.0, 1.0)

    # e degenerate (p3==p4): t=0, s = clip(-c/a)
    s[e_degenerate] = np.clip(-c[e_degenerate] / np.where(a[e_degenerate] > eps, a[e_degenerate], 1.0), 0.0, 1.0)

    # general case
    g = general
    safe_denom = np.where(denom[g] > eps, denom[g], 1.0)
    s_g = np.where(denom[g] > eps, np.clip((b[g] * f[g] - c[g] * e[g]) / safe_denom, 0.0, 1.0), 0.0)
    t_g = np.where(e[g] > eps, (b[g] * s_g + f[g]) / np.where(e[g] > eps, e[g], 1.0), 0.0)

    # clamp t into [0,1], re-derive s if t was clamped
    below = t_g < 0.0
    above = t_g > 1.0
    t_g = np.clip(t_g, 0.0, 1.0)
    s_g = np.where(
        below, np.clip(-c[g] / np.where(a[g] > eps, a[g], 1.0), 0.0, 1.0),
        np.where(above, np.clip((b[g] - c[g]) / np.where(a[g] > eps, a[g], 1.0), 0.0, 1.0), s_g)
    )

    s[g] = s_g
    t[g] = t_g

    closest1 = p1 + d1 * s[:, None]
    closest2 = p3 + d2 * t[:, None]
    dist = np.linalg.norm(closest1 - closest2, axis=1)
    # both-degenerate pairs: just point-point distance
    if np.any(both_degenerate):
        dist[both_degenerate] = np.linalg.norm(p1[both_degenerate] - p3[both_degenerate], axis=1)
    return dist


def _segment_segment_distance(p1, p2, p3, p4) -> float:
    """Closest distance between two 3D line segments. Standard
    closest-point-between-segments algorithm (Ericson, "Real-Time
    Collision Detection")."""
    d1 = p2 - p1
    d2 = p4 - p3
    r = p1 - p3
    a = np.dot(d1, d1)
    e = np.dot(d2, d2)
    f = np.dot(d2, r)
    if a <= 1e-12 and e <= 1e-12:
        return np.linalg.norm(p1 - p3)
    if a <= 1e-12:
        s = 0.0
        t = np.clip(f / e, 0.0, 1.0)
    else:
        c = np.dot(d1, r)
        if e <= 1e-12:
            t = 0.0
            s = np.clip(-c / a, 0.0, 1.0)
        else:
            b = np.dot(d1, d2)
            denom = a * e - b * b
            s = np.clip((b * f - c * e) / denom, 0.0, 1.0) if denom > 1e-12 else 0.0
            t = (b * s + f) / e
            if t < 0.0:
                t = 0.0
                s = np.clip(-c / a, 0.0, 1.0)
            elif t > 1.0:
                t = 1.0
                s = np.clip((b - c) / a, 0.0, 1.0)
    closest1 = p1 + d1 * s
    closest2 = p3 + d2 * t
    return float(np.linalg.norm(closest1 - closest2))

File: frontend/src/lib/kinematics.js
// Forward kinematics for all supported robots, mirroring the backend's DH-parameter
// math exactly (app/core/kinematics.py) so the 3D arm renders the same
// joint positions the solver actually computed -- this is purely a
// rendering helper; all solving happens server-side.

export const UR5_SPEC = {
  name: 'ur5',
  a: [0, -0.425, -0.39225, 0, 0, 0],
  d: [0.089159, 0, 0, 0.10915, 0.09465, 0.0823],
  alpha: [Math.PI / 2, 0, 0, Math.PI / 2, -Math.PI / 2, 0],
  linkRadius: [0.06, 0.05, 0.045, 0.04, 0.04, 0.035],
};

export const PLANAR3DOF_SPEC = {
  name: 'planar3dof',
  a: [0.4, 0.3, 0.2],
  d: [0, 0, 0],
  alpha: [0, 0, 0],
  linkRadius: [0.03, 0.025, 0.02],
};

export const FRANKA_PANDA_SPEC = {
  name: 'franka_panda',
  a: [0, 0, 0, 0.0825, -0.0825, 0, 0.088],
  d: [0.333, 0, 0.316, 0, 0.384, 0, 0.107],
  alpha: [0, -Math.PI / 2, Math.PI / 2, Math.PI / 2, -Math.PI / 2, Math.PI / 2, Math.PI / 2],
  linkRadius: [0.08, 0.07, 0.07, 0.06, 0.06, 0.05, 0.04],
};

export const ROBOT_SPECS = {
  planar3dof:   PLANAR3DOF_SPEC,
  ur5:          UR5_SPEC,
  franka_panda: FRANKA_PANDA_SPEC,
};

// Reasonable idle poses per robot (all within joint limits)
export const ROBOT_IDLE_Q = {
  planar3dof:   [0, 0, 0],
  ur5:          [0, -0.7, 0.9, -0.4, 0.6, 0],
  franka_panda: [0, -0.3, 0, -2.0, 0, 1.8, 0],  // q4 must be negative per Franka limits
};

function dhTransform(theta, d, a, alpha) {
  const ct = Math.cos(theta), st = Math.sin(theta);
  const ca = Math.cos(alpha), sa = Math.sin(alpha);
  // Row-major 4x4, matches the backend's _dh_transform exactly.
  return [
    [ct, -st * ca, st * sa, a * ct],
    [st, ct * ca, -ct * sa, a * st],
    [0, sa, ca, d],
    [0, 0, 0, 1],
  ];
}

function matMul(A, B) {
  const result = Array.from({ length: 4 }, () => [0, 0, 0, 0]);
  for (let i = 0; i < 4; i++) {
    for (let j = 0; j < 4; j++) {
      let sum = 0;
      for (let k = 0; k < 4; k++) sum += A[i][k] * B[k][j];
      result[i][j] = sum;
    }
  }
  return result;
}

const IDENTITY = [
  [1, 0, 0, 0],
  [0, 1, 0, 0],
  [0, 0, 1, 0],
  [0, 0, 0, 1],
];

/**
 * Returns an array of (n+1) {position: [x,y,z], matrix} for each joint
 * origin in base frame, given joint angles q (length n). Matches the
 * backend's forward_kinematics_chain exactly.
 */
export function forwardKinematicsChain(spec, q) {
  const n = spec.a.length;
  let T = IDENTITY;
  const chain = [{ position: [0, 0, 0], matrix: IDENTITY }];
  for (let i = 0; i < n; i++) {
    const Ti = dhTransform(q[i], spec.d[i], spec.a[i], spec.alpha[i]);
    T = matMul(T, Ti);
    chain.push({
      position: [T[0][3], T[1][3], T[2][3]],
      matrix: T,
    });
  }
  return chain;
}

/** Extracts a [x,y,z] quaternion-free rotation as a 3x3 for a chain entry. */
export function rotationOf(chainEntry) {
  const m = chainEntry.matrix;
  return [
    [m[0][0], m[0][1], m[0][2]],
    [m[1][0], m[1][1], m[1][2]],
    [m[2][0], m[2][1], m[2][2]],
  ];
}

/** 3x3 rotation matrix -> [x,y,z,w] quaternion (three.js convention). */
export function matrixToQuaternion(R) {
  const m00 = R[0][0], m01 = R[0][1], m02 = R[0][2];
  const m10 = R[1][0], m11 = R[1][1], m12 = R[1][2];
  const m20 = R[2][0], m21 = R[2][1], m22 = R[2][2];
  const trace = m00 + m11 + m22;
  let x, y, z, w;
  if (trace > 0) {
    const s = 0.5 / Math.sqrt(trace + 1.0);
    w = 0.25 / s;
    x = (m21 - m12) * s;
    y = (m02 - m20) * s;
    z = (m10 - m01) * s;
  } else if (m00 > m11 && m00 > m22) {
    const s = 2.0 * Math.sqrt(1.0 + m00 - m11 - m22);
    w = (m21 - m12) / s;
    x = 0.25 * s;
    y = (m01 + m10) / s;
    z = (m02 + m20) / s;
  } else if (m11 > m22) {
    const s = 2.0 * Math.sqrt(1.0 + m11 - m00 - m22);
    w = (m02 - m20) / s;
    x = (m01 + m10) / s;
    y = 0.25 * s;
    z = (m12 + m21) / s;
  } else {
    const s = 2.0 * Math.sqrt(1.0 + m22 - m00 - m11);
    w = (m10 - m01) / s;
    x = (m02 + m20) / s;
    y = (m12 + m21) / s;
    z = 0.25 * s;
  }
  return [x, y, z, w];
}

/** Minimum non-adjacent-link segment distance, mirrors the backend's
 * self_collision_min_distance for client-side collision-glow rendering
 * (visual only -- the backend's value is authoritative for metrics). */
export function selfCollisionMinDistance(spec, chainPositions) {
  const nLinks = spec.a.length;
  let minD = Infinity;
  for (let i = 0; i < nLinks - 1; i++) {
    for (let j = i + 2; j < nLinks; j++) {
      const d = segmentSegmentDistance(
        chainPositions[i], chainPositions[i + 1],
        chainPositions[j], chainPositions[j + 1]
      );
      const adjusted = d - (spec.linkRadius[i] + spec.linkRadius[j]);
      if (adjusted < minD) minD = adjusted;
    }
  }
  return Number.isFinite(minD) ? minD : 1.0;
}

function sub(a, b) { return [a[0] - b[0], a[1] - b[1], a[2] - b[2]]; }
function dot(a, b) { return a[0] * b[0] + a[1] * b[1] + a[2] * b[2]; }
function add(a, b, s) { return [a[0] + b[0] * s, a[1] + b[1] * s, a[2] + b[2] * s]; }
function norm(a) { return Math.sqrt(dot(a, a)); }

function segmentSegmentDistance(p1, p2, p3, p4) {
  const d1 = sub(p2, p1), d2 = sub(p4, p3), r = sub(p1, p3);
  const a = dot(d1, d1), e = dot(d2, d2), f = dot(d2, r);
  const eps = 1e-12;
  let s, t;
  if (a <= eps && e <= eps) return norm(sub(p1, p3));
  if (a <= eps) {
    s = 0; t = Math.min(Math.max(f / e, 0), 1);
  } else {
    const c = dot(d1, r);
    if (e <= eps) {
      t = 0; s = Math.min(Math.max(-c / a, 0), 1);
    } else {
      const b = dot(d1, d2);
      const denom = a * e - b * b;
      s = denom > eps ? Math.min(Math.max((b * f - c * e) / denom, 0), 1) : 0;
      t = (b * s + f) / e;
      if (t < 0) { t = 0; s = Math.min(Math.max(-c / a, 0), 1); }
      else if (t > 1) { t = 1; s = Math.min(Math.max((b - c) / a, 0), 1); }
    }
  }
  const c1 = add(p1, d1, s), c2 = add(p3, d2, t);
  return norm(sub(c1, c2));
}
